#include "ai.h"
#include "ui_ai.h"
#include <QMessageBox>
#include <QJsonObject>
#include <QJsonDocument>
#include <QPainter>
#include <QDateTime>
#include <QApplication>
#include <QMetaObject>
#include <QThread>
#include <QScrollBar>
#include <QColor>
#include <QPainterPath> // æ–°å¢ï¼šç”¨äºç»˜åˆ¶åœ†å½¢å¤´åƒ
// 1. æ°”æ³¡ç»˜åˆ¶ï¼šå®Œå…¨ä¿ç•™ä½ çš„ä»£ç 
void ChatDelegate::paint(QPainter *painter, const QStyleOptionViewItem &option, const QModelIndex &index) const {
    painter->save();
    painter->setRenderHints(QPainter::Antialiasing | QPainter::TextAntialiasing | QPainter::SmoothPixmapTransform); // æ–°å¢ï¼šå›¾ç‰‡å¹³æ»‘ç¼©æ”¾

    QString sender = index.data(SenderRole).toString();
    QString content = index.data(ContentRole).toString();
    qreal opacity = index.data(OpacityRole).toReal();
    if (opacity <= 0) opacity = 0.01;

    // åŸæœ‰æ–‡å­—å°ºå¯¸è®¡ç®—é€»è¾‘ä¸å˜
    QFont font = QApplication::font();
    QFontMetrics fm(font);
    QRect textRect = fm.boundingRect(QRect(0, 0, option.rect.width() * 0.75, 1000), Qt::TextWordWrap, content);
    QSize bubbleSize = QSize(textRect.width() + 28, textRect.height() + 20);

    // ğŸ”´ å¤´åƒé…ç½®ï¼ˆä»…AIæ°”æ³¡æ˜¾ç¤ºï¼‰
    const int avatarSize = 40; // å¤´åƒå°ºå¯¸ï¼ˆ40x40ï¼Œå¯è°ƒæ•´ï¼‰
    const int avatarSpacing = 10; // å¤´åƒä¸æ°”æ³¡çš„é—´è·
    QRect avatarRect; // å¤´åƒç»˜åˆ¶åŒºåŸŸ
    QPixmap aiAvatar; // AIå¤´åƒå›¾ç‰‡

    QRect bubbleRect;
    QLinearGradient bubbleGradient;
    if (sender == "ç”¨æˆ·") {
        // ç”¨æˆ·æ°”æ³¡ï¼šåŸæœ‰é€»è¾‘ä¸å˜ï¼ˆæ— å¤´åƒï¼‰
        bubbleRect = QRect(option.rect.right() - bubbleSize.width(), option.rect.top() + 10, bubbleSize.width(), bubbleSize.height());
        bubbleGradient = QLinearGradient(bubbleRect.topLeft(), bubbleRect.bottomRight());
        bubbleGradient.setColorAt(0, QColor(66, 153, 225, int(opacity * 255)));
        bubbleGradient.setColorAt(1, QColor(125, 184, 255, int(opacity * 255)));
    } else {
        // ğŸ”´ AIæ°”æ³¡ï¼šå…ˆç»˜åˆ¶å¤´åƒï¼Œå†è°ƒæ•´æ°”æ³¡ä½ç½®
        // 1. åŠ è½½AIå¤´åƒï¼ˆèµ„æºæ–‡ä»¶è·¯å¾„ï¼Œç¡®ä¿res.qrcé…ç½®æ­£ç¡®ï¼‰
        static QPixmap cachedAvatar; // é™æ€ç¼“å­˜ï¼Œé¿å…æ¯æ¬¡ç»˜åˆ¶éƒ½åŠ è½½
        if (cachedAvatar.isNull()) {
            // ä»èµ„æºæ–‡ä»¶åŠ è½½å›¾ç‰‡ï¼ˆè·¯å¾„å¯¹åº”res.qrcä¸­çš„é…ç½®ï¼‰
            cachedAvatar = QPixmap(":/resources/images/ai_avatar.png").scaled(
                avatarSize, avatarSize,
                Qt::KeepAspectRatio,
                Qt::SmoothTransformation // å¹³æ»‘ç¼©æ”¾
                );
        }
        aiAvatar = cachedAvatar;

        // 2. ç¡®å®šå¤´åƒä½ç½®ï¼ˆå·¦ä¾§å±…ä¸­å¯¹é½æ°”æ³¡ï¼‰
        avatarRect = QRect(
            option.rect.left() + 10, // å¤´åƒå·¦ä¾§è·ç¦»çª—å£å·¦è¾¹10px
            option.rect.top() + (bubbleSize.height() - avatarSize) / 2 + 10, // å‚ç›´å±…ä¸­ï¼ˆä¸æ°”æ³¡å¯¹é½ï¼‰
            avatarSize,
            avatarSize
            );

        // 3. ç»˜åˆ¶åœ†å½¢å¤´åƒï¼ˆè£å‰ªä¸ºåœ†å½¢ï¼ŒæŠ—é”¯é½¿ï¼‰
        painter->save();
        QPainterPath avatarPath;
        avatarPath.addEllipse(avatarRect); // åœ†å½¢è·¯å¾„
        painter->setClipPath(avatarPath); // è£å‰ªä¸ºåœ†å½¢

        // ç»˜åˆ¶å¤´åƒï¼ˆåŠ è½½å¤±è´¥æ—¶æ˜¾ç¤ºé»˜è®¤ç°è‰²åœ†å½¢ï¼‰
        if (!aiAvatar.isNull()) {
            painter->drawPixmap(avatarRect, aiAvatar);
        } else {
            painter->setBrush(QColor(220, 220, 220));
            painter->drawEllipse(avatarRect);
        }

        // ç»˜åˆ¶å¤´åƒè¾¹æ¡†ï¼ˆå¯é€‰ï¼šç™½è‰²ç»†è¾¹æ¡†ï¼Œå¢å¼ºç«‹ä½“æ„Ÿï¼‰
        painter->setClipRect(option.rect); // å–æ¶ˆè£å‰ª
        painter->setBrush(Qt::NoBrush);
        painter->setPen(QPen(QColor(255, 255, 255), 2));// ç™½è‰²è¾¹æ¡†ï¼Œ2pxå®½
        painter->drawEllipse(avatarRect);
        painter->restore();

        // 4. è°ƒæ•´AIæ°”æ³¡ä½ç½®ï¼šå‘å³åç§»ï¼ˆå¤´åƒå®½åº¦+é—´è·ï¼‰ï¼Œé¿å…é‡å 
        bubbleRect = QRect(
            option.rect.left() + avatarSize + avatarSpacing + 10, // åŸæœ‰10px + å¤´åƒ40px + é—´è·10px
            option.rect.top() + 10,
            bubbleSize.width(),
            bubbleSize.height()
            );

        // æ°”æ³¡æ¸å˜ï¼ˆåŸæœ‰é€»è¾‘ä¸å˜ï¼‰
        bubbleGradient = QLinearGradient(bubbleRect.topLeft(), bubbleRect.bottomRight());
        bubbleGradient.setColorAt(0, QColor(240, 240, 240, int(opacity * 255)));
        bubbleGradient.setColorAt(1, QColor(224, 224, 224, int(opacity * 255)));
    }

    // ç»˜åˆ¶æ°”æ³¡ï¼ˆåŸæœ‰é€»è¾‘ä¸å˜ï¼‰
    painter->setBrush(bubbleGradient);
    painter->setPen(Qt::NoPen);
    painter->drawRoundedRect(bubbleRect, 20, 20);

    // ç»˜åˆ¶æ–‡å­—ï¼ˆåŸæœ‰é€»è¾‘ä¸å˜ï¼‰
    painter->setPen(QColor(45, 55, 72, int(opacity * 255)));
    painter->setFont(font);
    painter->drawText(QRect(bubbleRect.x() + 14, bubbleRect.y() + 10, bubbleRect.width() - 28, bubbleRect.height() - 20),
                      Qt::TextWordWrap, content);

    painter->restore();
}
// æ–°å¢ï¼šæ›´æ–°æ°”æ³¡é€æ˜åº¦ï¼Œå®ç°æ·¡å…¥åŠ¨ç”»
void AIQueryWidget::paintEvent(QPaintEvent *event) {
    Q_UNUSED(event);
    QPainter painter(this);
    // æ¸å˜é…ç½®ï¼šä»æµ…è“åˆ°æµ…ç´«ï¼ˆå¯è‡ªè¡Œè°ƒæ•´é¢œè‰²ï¼‰
    QLinearGradient gradient(0, 0, this->width(), this->height());
    gradient.setColorAt(0, QColor(245, 248, 255)); // å·¦ä¸Šè§’é¢œè‰²ï¼ˆæ·¡è“ï¼‰
    gradient.setColorAt(1, QColor(250, 245, 255)); // å³ä¸‹è§’é¢œè‰²ï¼ˆæ·¡ç´«ï¼‰
    // ç»˜åˆ¶èƒŒæ™¯
    painter.setBrush(gradient);
    painter.setPen(Qt::NoPen);
    painter.drawRect(this->rect());
}

void AIQueryWidget::updateItemOpacity(QListWidgetItem *item) {
    if (!item || !ui->chatListWidget) return;

    qreal opacity = item->data(ChatDelegate::OpacityRole).toReal();
    if (opacity >= 1.0) {
        return;
    }

    opacity += 0.08;
    if (opacity > 1.0) opacity = 1.0;
    item->setData(ChatDelegate::OpacityRole, opacity);

    int interval = 50;

    int row = ui->chatListWidget->row(item);
    if (row != -1) {
        QModelIndex index = ui->chatListWidget->model()->index(row, 0);
        ui->chatListWidget->update(index);
    }

    QTimer::singleShot(interval, this, [=]() {
        updateItemOpacity(item);
    });
}

// 2. Itemå°ºå¯¸è®¡ç®—ï¼šå®Œå…¨ä¿ç•™ä½ çš„ä»£ç 
QSize ChatDelegate::sizeHint(const QStyleOptionViewItem &option, const QModelIndex &index) const {
    QString content = index.data(ContentRole).toString();
    QFontMetrics fm(QApplication::font());
    QRect textRect = fm.boundingRect(
        QRect(0, 0, option.rect.width() * 0.75, 1000),
        Qt::TextWordWrap,
        content
        );
    return QSize(option.rect.width(), textRect.height() + 50);
}

// æ–°å¢ï¼š3. å®‰å…¨åˆ é™¤"æ­£åœ¨æ€è€ƒ"Itemï¼ˆå¼ºåˆ¶ä¸»çº¿ç¨‹ï¼‰
void AIQueryWidget::safeRemoveLoadingItem() {
    // åŒé‡ç©ºæŒ‡é’ˆåˆ¤æ–­ï¼Œæœç»é‡æŒ‡é’ˆ
    if (this->thread() != QThread::currentThread()) {
        // ä¸åœ¨ä¸»çº¿ç¨‹ï¼ŒæŠ•é€’ç»™ä¸»çº¿ç¨‹
        QMetaObject::invokeMethod(this, &AIQueryWidget::safeRemoveLoadingItem, Qt::QueuedConnection);
        return;
    }

    if (m_loadingItem && ui->chatListWidget->itemWidget(m_loadingItem) == nullptr) {
        // ç¡®è®¤Itemå±äºå½“å‰ListWidgetï¼Œå†åˆ é™¤
        int index = ui->chatListWidget->row(m_loadingItem);
        if (index != -1) {
            delete ui->chatListWidget->takeItem(index); // å…ˆä»ListWidgetç§»é™¤ï¼Œå†åˆ é™¤
        }
        m_loadingItem = nullptr; // ç½®ç©ºï¼Œé¿å…é‡æŒ‡é’ˆ
    }
}

// æ–°å¢ï¼š4. å®‰å…¨æ·»åŠ èŠå¤©Itemï¼ˆå¼ºåˆ¶ä¸»çº¿ç¨‹ï¼‰
// ä¿®æ”¹ï¼šåœ¨å®‰å…¨æ·»åŠ Itemæ—¶ï¼Œåˆå§‹åŒ–é€æ˜åº¦å¹¶å¯åŠ¨æ·¡å…¥åŠ¨ç”»
void AIQueryWidget::safeAddChatItem(const QString &sender, const QString &content) {
    if (this->thread() != QThread::currentThread()) {
        QMetaObject::invokeMethod(this, [this, sender, content]() {
            this->safeAddChatItem(sender, content);
        }, Qt::QueuedConnection);
        return;
    }

    if (!ui->chatListWidget) return;

    QListWidgetItem *item = new QListWidgetItem(ui->chatListWidget);
    item->setData(ChatDelegate::SenderRole, sender);
    item->setData(ChatDelegate::ContentRole, content);
    // ğŸ”´ åˆå§‹åŒ–é€æ˜åº¦ä¸º0.0ï¼ˆæ·¡å…¥åŠ¨ç”»èµ·ç‚¹ï¼šå®Œå…¨é€æ˜ï¼‰
    item->setData(ChatDelegate::OpacityRole, 0.0);
    ui->chatListWidget->addItem(item);
    ui->chatListWidget->update();

    // ğŸ”´ å¯åŠ¨æ·¡å…¥åŠ¨ç”»ï¼ˆå»¶è¿Ÿ10msï¼Œé¿å…ä¸UIåˆ·æ–°å†²çªï¼‰
    QTimer::singleShot(10, this, [=]() {
        updateItemOpacity(item);
    });

    safeScrollToBottom(); // åŸæœ‰æ»šåŠ¨é€»è¾‘ä¸å˜
}

// æ–°å¢ï¼š5. å®‰å…¨æ»šåŠ¨åˆ°åº•éƒ¨ï¼ˆå¼ºåˆ¶ä¸»çº¿ç¨‹ï¼‰
void AIQueryWidget::safeScrollToBottom() {
    if (this->thread() != QThread::currentThread()) {
        QMetaObject::invokeMethod(this, &AIQueryWidget::safeScrollToBottom, Qt::QueuedConnection);
        return;
    }

    if (ui->chatListWidget) {
        int count = ui->chatListWidget->count();
        if (count > 0) {
            ui->chatListWidget->scrollToItem(ui->chatListWidget->item(count - 1), QAbstractItemView::PositionAtBottom);
        }
    }
}

// 6. å–æ¶ˆå½“å‰è¯·æ±‚ï¼ˆå¼ºåŒ–èµ„æºé‡Šæ”¾ï¼‰
void AIQueryWidget::cancelCurrentRequest() {
    // ç½‘ç»œè¯·æ±‚é‡Šæ”¾
    if (m_currentReply) {
        m_currentReply->disconnect();
        m_currentReply->abort();
        m_currentReply->deleteLater();
        m_currentReply = nullptr;
    }

    if (m_requestTimer) {
        m_requestTimer->stop();
    }

    // å®‰å…¨åˆ é™¤â€œæ­£åœ¨æ€è€ƒâ€Item
    safeRemoveLoadingItem();

    // ğŸ”´ å¼ºåˆ¶é‡ç½®è¯·æ±‚çŠ¶æ€
    m_isRequesting = false;
}
// 7. è¶…æ—¶å¤„ç†ï¼ˆå®‰å…¨æç¤ºï¼‰
void AIQueryWidget::onRequestTimeout() {
    cancelCurrentRequest();
    safeAddChatItem("AI", "æŸ¥è¯¢è¶…æ—¶ï¼è¯·æ£€æŸ¥ç½‘ç»œæˆ–ç¨åé‡è¯•~");
}

// 8. çª—å£åˆå§‹åŒ–ï¼ˆå¼ºåŒ–æ§ä»¶å®‰å…¨ï¼‰
AIQueryWidget::AIQueryWidget(QWidget *parent) :
    QWidget(parent),
    ui(new Ui::AIQueryWidget),
    m_netManager(new QNetworkAccessManager(this)),
    m_apiUrl("http://localhost:8000/query_flight"),
    m_chatDelegate(new ChatDelegate(this))
{
    ui->setupUi(this);
    this->setWindowTitle("AI assistant");
    ui->queryButton->setStyleSheet(R"(
        QPushButton {
            color: black;           /* æŒ‰é’®å­—ä½“é»‘è‰² */
            background-color: #f0f0f0; /* æŒ‰é’®èƒŒæ™¯æµ…ç°ï¼ˆå¯é€‰ï¼‰ */
            border: 1px solid #ccc;    /* æŒ‰é’®è¾¹æ¡†ï¼ˆå¯é€‰ï¼‰ */
            padding: 5px 15px;         /* æŒ‰é’®å†…è¾¹è·ï¼ˆå¯é€‰ï¼‰ */
        }
        QPushButton:hover { /* é¼ æ ‡æ‚¬æµ®æ ·å¼ï¼ˆå¯é€‰ï¼‰ */
            background-color: #e0e0e0;
        }
        QPushButton:pressed { /* ç‚¹å‡»æ ·å¼ï¼ˆå¯é€‰ï¼‰ */
            background-color: #d0d0d0;
        }
    )");
    // ä½ çš„åŸå§‹é…ç½®ï¼Œæ·»åŠ åˆ·æ–°ç¡®ä¿åˆå§‹åŒ–æ­£å¸¸
    ui->chatListWidget->setItemDelegate(m_chatDelegate);
    ui->chatListWidget->setSpacing(10);
    ui->chatListWidget->setVerticalScrollMode(QAbstractItemView::ScrollPerPixel);
    QScrollBar *vScrollBar = ui->chatListWidget->verticalScrollBar();
    if (vScrollBar) {
        vScrollBar->setSingleStep(10);  // é¼ æ ‡æ»šè½®/ç®­å¤´é”®å•æ¬¡æ»šåŠ¨10åƒç´ ï¼ˆè¶Šå°è¶Šç»†è…»ï¼‰
        vScrollBar->setPageStep(80);    // ç‚¹å‡»æ»šåŠ¨æ¡ç©ºç™½å¤„å•æ¬¡æ»šåŠ¨80åƒç´ ï¼ˆé€‚é…å±å¹•é«˜åº¦ï¼‰
    }
    ui->chatListWidget->setSizePolicy(QSizePolicy::Expanding, QSizePolicy::Expanding);
    ui->chatListWidget->update(); // åˆå§‹åŒ–ååˆ·æ–°
    ui->chatListWidget->setStyleSheet("background-color: transparent; border: none;");
    // å¯é€‰ï¼šè®¾ç½®è¾“å…¥æ¡†å’ŒæŒ‰é’®åŒºåŸŸèƒŒæ™¯é€æ˜ï¼ˆä¿æŒé£æ ¼ç»Ÿä¸€ï¼‰
    ui->questionLineEdit->setStyleSheet(R"(
        QLineEdit, QTextEdit, QPlainTextEdit {
            color: black!important;       /* è¾“å…¥å­—ä½“é»‘è‰² */
            background-color: white; /* èƒŒæ™¯ç™½è‰²ï¼ˆå¯é€‰ï¼Œé¿å…å…¨å±€èƒŒæ™¯ç™½ï¼‰ */
            border: 1px solid #ccc;  /* è¾¹æ¡†ï¼ˆå¯é€‰ï¼Œæ›´æ¸…æ™°ï¼‰ */
            padding: 5px;            /* å†…è¾¹è·ï¼ˆå¯é€‰ï¼‰ */
        }
        QLineEdit:focus { /* èšç„¦æ—¶æ ·å¼ï¼ˆå¯é€‰ï¼‰ */
            border-color: #66afe9;
            outline: none;
        }
    )");
    // è¶…æ—¶è®¡æ—¶å™¨ï¼ˆå¼ºåŒ–ç©ºæŒ‡é’ˆåˆ¤æ–­ï¼‰
    m_requestTimer = new QTimer(this);
    m_requestTimer->setSingleShot(true);
    m_requestTimer->setInterval(30000);
    connect(m_requestTimer, &QTimer::timeout, this, &AIQueryWidget::onRequestTimeout);

    // åˆå§‹æ¬¢è¿æ¶ˆæ¯ï¼ˆå®‰å…¨æ·»åŠ ï¼‰
    safeAddChatItem("AI", "æˆ‘æ˜¯ä¸€ä¸ªèƒ½å›ç­”å…³äºèˆªç­æŸ¥è¯¢çš„é—®é¢˜çš„åŠ©æ‰‹ã€‚å¦‚æœæ‚¨æœ‰å…³äºèˆªç­çš„é—®é¢˜ï¼Œè¯·å‘Šè¯‰æˆ‘å…·ä½“çš„å‡ºå‘åœ°ã€ç›®çš„åœ°å’Œæ—¥æœŸã€‚");

    // ä¿¡å·è¿æ¥ï¼ˆæ–­å¼€æ—§è¿æ¥ï¼Œé¿å…é‡å¤è¿æ¥ï¼‰
    disconnect(m_netManager, &QNetworkAccessManager::finished, this, &AIQueryWidget::onReplyFinished);
    connect(m_netManager, &QNetworkAccessManager::finished, this, &AIQueryWidget::onReplyFinished);

    disconnect(ui->questionLineEdit, &QLineEdit::returnPressed, this, &AIQueryWidget::onReturnPressed);
    connect(ui->questionLineEdit, &QLineEdit::returnPressed, this, &AIQueryWidget::onReturnPressed);
}

// 9. ææ„å‡½æ•°ï¼ˆå½»åº•é‡Šæ”¾UIèµ„æºï¼‰
AIQueryWidget::~AIQueryWidget() {
    // é¡ºåº1ï¼šå…ˆåœæ­¢è®¡æ—¶å™¨ï¼ˆå­èµ„æºï¼‰
    if (m_requestTimer) {
        m_requestTimer->stop();
        m_requestTimer->deleteLater();
        m_requestTimer = nullptr;
    }

    // é¡ºåº2ï¼šå–æ¶ˆç½‘ç»œè¯·æ±‚ï¼ˆå­èµ„æºï¼‰
    if (m_currentReply) {
        m_currentReply->abort();
        m_currentReply->disconnect();
        m_currentReply->deleteLater();
        m_currentReply = nullptr; // æ™®é€šæŒ‡é’ˆç½®ç©º
    }

    // é¡ºåº3ï¼šé‡Šæ”¾èŠå¤©Itemï¼ˆå­èµ„æºï¼Œä¾èµ–chatListWidgetï¼‰
    if (ui && ui->chatListWidget) {
        // ç»ˆææ–¹æ¡ˆï¼šå¾ªç¯åˆ é™¤ç¬¬ä¸€ä¸ªItemï¼Œç›´åˆ°æ²¡æœ‰Itemä¸ºæ­¢ï¼ˆé€‚é…æ‰€æœ‰Qtç‰ˆæœ¬ï¼‰
        while (true) {
            QListWidgetItem* item = ui->chatListWidget->item(0); // è·å–ç¬¬ä¸€ä¸ªItem
            if (!item) break; // æ²¡æœ‰Itemäº†ï¼Œé€€å‡ºå¾ªç¯
            delete item; // åˆ é™¤Item
        }
        ui->chatListWidget->clear(); // æœ€åæ¸…ç©ºåˆ—è¡¨
    }

    // é¡ºåº4ï¼šé‡Šæ”¾ç½‘ç»œç®¡ç†å™¨ï¼ˆå­èµ„æºï¼‰
    if (m_netManager) {
        m_netManager->disconnect();
        m_netManager->deleteLater();
        m_netManager = nullptr;
    }

    // é¡ºåº5ï¼šé‡Šæ”¾å§”æ‰˜ï¼ˆå­èµ„æºï¼‰
    delete m_chatDelegate;
    m_chatDelegate = nullptr;

    // é¡ºåº6ï¼šæœ€åé‡Šæ”¾UIï¼ˆçˆ¶èµ„æºï¼ŒåŒ…å«chatListWidgetï¼‰
    delete ui;
    ui = nullptr;
}

// 10. ç‚¹å‡»æŸ¥è¯¢æŒ‰é’®ï¼ˆå¼ºåŒ–å®‰å…¨åˆ¤æ–­ï¼‰
void AIQueryWidget::on_queryButton_clicked() {
    // é˜²æ­¢çª—å£æ§ä»¶å·²é”€æ¯
    if (!ui->questionLineEdit || !ui->chatListWidget) return;

    QString question = ui->questionLineEdit->text().trimmed();
    if (question.isEmpty()) {
        QMessageBox::warning(this, "è­¦å‘Š", "è¯·è¾“å…¥æŸ¥è¯¢é—®é¢˜ï¼");
        return;
    }

    // ğŸ”´ å…ˆæ·»åŠ ç”¨æˆ·æ°”æ³¡ï¼ˆæ— è®ºè¯·æ±‚æ˜¯å¦èƒ½å‘ï¼Œå…ˆæ˜¾ç¤ºç”¨æˆ·è¾“å…¥ï¼‰
    safeAddChatItem("ç”¨æˆ·", question);

    // å†æ£€æŸ¥è¯·æ±‚çŠ¶æ€
    if (m_isRequesting) {
        QMessageBox::information(this, "æç¤º", "æ­£åœ¨å¤„ç†æ‚¨çš„è¯·æ±‚ï¼Œè¯·ç¨å€™~");
        ui->questionLineEdit->clear(); // æ¸…ç©ºè¾“å…¥æ¡†
        return;
    }
    m_isRequesting = true; // æ ‡è®°ä¸ºæ­£åœ¨å¤„ç†

    // æ·»åŠ â€œæ­£åœ¨æ€è€ƒâ€Item
    QMetaObject::invokeMethod(this, [=]() {
        m_loadingItem = new QListWidgetItem(ui->chatListWidget);
        m_loadingItem->setData(ChatDelegate::SenderRole, "AI");
        m_loadingItem->setData(ChatDelegate::ContentRole, "æ­£åœ¨æ€è€ƒ.........");
        ui->chatListWidget->addItem(m_loadingItem);
        ui->chatListWidget->update();
        safeScrollToBottom();
    }, Qt::QueuedConnection);

    // ğŸ”´ ç¡®ä¿å‘é€è¯·æ±‚çš„ä»£ç è¢«æ‰§è¡Œï¼ˆåç«¯æ²¡æ”¶åˆ°è¯·æ±‚çš„æ ¸å¿ƒåŸå› ï¼‰
    QJsonObject jsonObj;
    jsonObj["question"] = question;
    QJsonDocument jsonDoc(jsonObj);
    QByteArray postData = jsonDoc.toJson();

    QNetworkRequest request;
    request.setUrl(QUrl(m_apiUrl));
    request.setHeader(QNetworkRequest::ContentTypeHeader, "application/json");

    // é‡ç½®å½“å‰è¯·æ±‚ï¼ˆé¿å…æ—§è¯·æ±‚å¹²æ‰°ï¼‰
    if (m_currentReply) {
        m_currentReply->disconnect();
        m_currentReply->deleteLater();
    }
    m_currentReply = m_netManager->post(request, postData);
    m_requestTimer->start(); // å¯åŠ¨è¶…æ—¶è®¡æ—¶å™¨

    ui->questionLineEdit->clear();
}

// 11. ç½‘ç»œå“åº”å¤„ç†ï¼ˆæ‰€æœ‰UIæ“ä½œèµ°å®‰å…¨å‡½æ•°ï¼‰
void AIQueryWidget::onReplyFinished(QNetworkReply *reply) {


    // ğŸ”´ ç¬¬äºŒæ­¥ï¼šæ£€æŸ¥å…³é”®æ§ä»¶æ˜¯å¦å­˜åœ¨
    if (!ui || !ui->chatListWidget || !ui->queryButton) {
        reply->deleteLater();
        return;
    }
    if (m_requestTimer) {
        m_requestTimer->stop();
    }

    // å®‰å…¨åˆ é™¤â€œæ­£åœ¨æ€è€ƒâ€Item
    safeRemoveLoadingItem();

    // å¤„ç†å“åº”
    if (reply->error() != QNetworkReply::NoError) {
        if (reply->error() != QNetworkReply::OperationCanceledError) {
            safeAddChatItem("AI", QString("æŸ¥è¯¢å¤±è´¥ï¼š%1").arg(reply->errorString()));
        }
    } else {
        QByteArray response = reply->readAll();
        QJsonDocument jsonDoc = QJsonDocument::fromJson(response);
        if (!jsonDoc.isObject()) {
            safeAddChatItem("AI", "æŸ¥è¯¢å¤±è´¥ï¼šæ— æ•ˆå“åº”");
        } else {
            QString answer = jsonDoc.object()["answer"].toString();
            safeAddChatItem("AI", answer);
        }
    }

    // ğŸ”´ å¼ºåˆ¶é‡ç½®è¯·æ±‚çŠ¶æ€
    m_currentReply = nullptr;
    m_isRequesting = false; // æ— è®ºæˆåŠŸå¤±è´¥ï¼Œéƒ½é‡ç½®ä¸ºfalse

    // é‡Šæ”¾reply
    reply->disconnect();
    reply->deleteLater();
}
// 12. å›è½¦è§¦å‘ï¼šä¿ç•™ä½ çš„ä»£ç 
void AIQueryWidget::onReturnPressed() {
    on_queryButton_clicked();
}
